### DISCUSSION TITLE
hot-replace build / build with symlink

### DISCUSSION THREAD

**Discussion by Suven** [withastro/roadmap#758] [Proposal]
### Body

# Summary

I want to be able to reexecute `npm run build` without downtimes. So if a dist-folder is already present, I would want build to build to somewhere else, and then swap those folders, if/when building succeeded.

# Background & Motivation

I want to use Astro for a static site but still allow CMS-Changes via statamic. I have no fast rebuild-requirements. For my needs it would be enough to rebuild every few minutes (maybe via cron). I am also not using a CDN / hoster like vercel which would me allow to do that easily. I am using classic bare metal webservers.

# Goals

- Keep the current build-experience as is, for fresh builds
- Make incremental builds 0-config swapping. 

# Example

- Instead of building to the desired directory, `run build` builds into `.$target`
- When finished, it runs `rm -rf ${target} && mv ".$target" $target`

**Comment by delucis** [comment]
`astro build` does support an `--outDir` flag, so maybe that helps to do this already?


I’d probably opt for something like this in your `package.json`:
```json
{
  "scripts": {
    "build": "astro build --outDir _temp && rm -rf dist && mv _temp dist"
  }
}
```

It’s not 100% zero downtime I guess though because there is a brief gap between files being deleted and replaced. Depends on your requirements.

**Reply by jase-languasco** [reply]
Instead of deleting and moving the new directory you can copy the contents into the build directory and then delete the temporary directory

`astro build --outDir _temp && cp -r _temp/* dist && rm -rf _tmp`

**Reply by delucis** [reply]
That won’t remove old assets from `dist/` that are no longer needed though, so depends a bit on if that’s acceptable or not. Could mean outdated files accumulate over time.

**Reply by Suven** [reply]
Rsync supports that. One could do something like `rsync -av --progress --checksum --delete ./dist/client/ ./serv`. It's also a bit faster as it skips identical files.

I'd still argue that not temporarily destroying the dist-folder while (re)building would be a nice default behaviour.

**Reply by monaye** [reply]
try using the symbolic instead of dis folder. so `dist` look at the build output folder. 
```
date=$(date +%Y%m%d%H%M%S) && astro build --outDir build/$date && ln -sfn build/$date dist
```
this will create a new folder for each build and override the symbolic link.
